### PR TITLE
Filter 13C isotopes

### PR DESCRIPTION
**Pull request by robinschmid**
Added a more comprehensive 13C isotope filter to the rows filter. Previously, we only filter for minimum signals in an isotope pattern. For the Isotope finder, which is more inclusive, we need a better 13C filter.
- for each row's m/z we search for the +1 13C isotope peak in the isotope pattern and score the height. 
- there might be multiple signals in range so all are considered - only one needs to match the height requirements which derived from estimated min / max number of carbons.
- estimates the minimum and maximum number of 13 carbons based on mass defect and m/z 
- optionally looks for a -1 13C peak in the pattern (with matching height) to flag the row as an isotope and filter it
- optionally uses other elements to make sure we are not looking at a 18O or any other isotope signal (searching for the main signal at -X)
- added plenty of tests to check the performance with multiple charge states considered etc

image

### CODE REVIEW

**Comment by robinschmid** [comment]
### Cummulative histogram of library matches that match isotope filters (13C = Isotope finder; 2iso = Isotope grouper)
image

image



### Cummulative spectral library matches matching the 13C filter (Isotope finder) or the 2 isotopes by isotope grouper
image

image

**Comment by SteffenHeu** [comment]
What exactly are the graphs meant to mean? I dont see the colour of no match and match 2iso in the charts although theyre present in the legend

**Comment by robinschmid** [comment]
Oh sorry forgot to say that.
So the no match was filtered. So I used 28 samples, each a mix of extracted compounds. I know what compounds to expect in which sample and filter for library matches to compounds that we expect in these samples. I removed unmatched from because I only wanted to look at those high quality matches.

match 2iso has no unique matches therefore its empty

**Comment by robinschmid** [comment]
I updated the figures there was something wrong with my plotting script. Now it makes more sense
image

File: src/main/java/io/github/mzmine/util/IntRange.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+package io.github.mzmine.util;
+
+/**
+ * @author Robin Schmid (https://github.com/robinschmid)
+ */
+public record IntRange(int minInclusive, int maxExclusive) {
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/util/IntRange.java:24]: Whats the benefit over the guava range?

**Comment by SteffenHeu** [comment]
Maybe add it as a parameter to the spectral library matching?

**Comment by robinschmid** [comment]
What do you mean by adding to as a parameter to spectral library matching?
This is a filter to not really a scoring method

File: src/main/java/io/github/mzmine/util/IntRange.java (hunk 1 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/util/IntRange.java:24]: what is the benefit of Range<Integer>.openClosed(min, max) over new IntRange(min, max) ?

Comment (robinschmid) [reply src/main/java/io/github/mzmine/util/IntRange.java:24]: primitives, record type, fixed behavior without all this open / closed stuff - the standard behavior of every index range

**Comment by robinschmid** [comment]
Tested and ready from my side @SteffenHeu

**Pull request merged**